Added logging to n8n

krynble opened the pull request:
This commit adds logging to n8n using the Winston library.

For now, this commit only allows logging to console (default behavior)
or file (need to pass in config via environment variables).

Other logging methods can be further implemented using hooks. These were
skipped for now as it would require adding more dependencies.

Logging level is `info`by default, meaning no additional messages would
be displayed at the moment. Logging level can be set to verbose or debug as
well to enrich the generated logs.

The ILogger interface was added to the workflow project as it would make
it available for all other projects but the implementation was done on
the cli project.

krynble committed:
Added logging to n8n

This commit adds logging to n8n using the Winston library.

For now, this commit only allows logging to console (default behavior)
or file (need to pass in config via environment variables).

Other logging methods can be further implemented using hooks. These were
skipped for now as it would require adding more dependencies.

Logging level is notice by default, meaning no additional messages would
be displayed at the moment. Logging level can be set to info or debug as
well to enrich the generated logs.

The ILogger interface was added to the workflow project as it would make
it available for all other projects but the implementation was done on
the cli project.

krynble committed:
Lint fixes and logging level naming. Also fixed the way we use the logger as it was not working previously

janober commented on packages/cli/src/ActiveWorkflowRunner.ts:
```diff
@@ -58,21 +59,27 @@ export class ActiveWorkflowRunner {
 
 		this.activeWorkflows = new ActiveWorkflows();
 
+		const logger = (global as any).logger as ILogger;  // tslint:disable-line:no-any
```
Can we not do that 1x outside of the class so that we do not have this line multiple times in the same file and can simply use directly `logger` everywhere inside of it?

krynble committed:
Improvements to logging framework

Using appropriate single quotes
Improving the way the logger is declared

ivov commented on packages/workflow/src/Interfaces.ts:
```diff
@@ -760,7 +760,7 @@ export interface IWorkflowSettings {
 }
 
-export type ILogTypes = "debug" | "verbose" | "info" | "warn" | "error";
+export type ILogTypes = 'debug' | 'verbose' | 'info' | 'warn' | 'error';
```
Or `LogTypes` since this is a union type, not an interface.

krynble committed:
Improved naming for Log Types

krynble committed:
Removed logger global variable, replacing it by a proxy

krynble committed:
Resolving conflicts with master branch

MedAliMarz committed:
Add logging to CLI commands

MedAliMarz committed:
Remove unused GenericHelpers

krynble committed:
Changed back some messages to console instead of logger and added npm
shortcuts for worker and webhook

krynble commented:
Thank you for your contributions @MedAliMarz 

I had to revert back some of the messages from the logger to console as some information is only relevant to the terminal output and have little / no relevance to be sent to a logger system (files, elasticsearch, loggly, etc.).

Also I added shortcuts for the `webhook` and `worker` commands @janober Let me know if you see any issues with this.

mutdmour commented on packages/cli/commands/execute.ts:
```diff
@@ -163,24 +172,26 @@ export class Execute extends Command {
-				this.log('Execution was NOT successfull:');
-				this.log('====================================');
-				this.log(JSON.stringify(data, null, 2));
+				console.info('Execution was NOT successfull. See log message for details.');
```
nitpick - typo. 
```suggestion
				console.info('Execution was NOT successful. See log message for details.');
```

krynble replied:
Thanks, will merge all suggestions into a single commit =)

mutdmour commented on packages/cli/commands/export/credentials.ts:
```diff
@@ -14,10 +14,17 @@ import {
 
 import {
 	Db,
-	GenericHelpers,
```
it looks like "logOutput" function from this file is not used anymore so maybe we should delete that

krynble replied:
Thanks, removed!

ahsanv commented on packages/cli/commands/import/credentials.ts:
```diff
@@ -37,17 +44,20 @@ export class ImportCredentialsCommand extends Command {
 			if (fs.existsSync(flags.input)) {
 				if (!fs.lstatSync(flags.input).isDirectory()) {
-					GenericHelpers.logOutput(`The paramenter --input must be a directory`);
+					console.info(`The paramenter --input must be a directory`);
```
instead of using `console.log`, why don't we use Logger for all logs, and setup transports according to log level? Say info and lower (more important) go to file and standard output, and higher than info (debug etc) go to only file?

So that we have a consistent implementation?

krynble replied:
Hey Ahsan.

This was a dillema during development. In some cases, I only want messages to be output to terminal only (they don't really make sens to be added to a log file as this is an import credentials command).

So I was using both the console.* messages with LoggerProxy.* messages according to the type of message.

For a log file, my expectations as a developer is to always have all errors and even better, have debug messages on log files to help me solve issues. So actually log files should contain all relevant logs and UI prompts (even if the interface is the terminal) should not be there.

Does this make sense? Let me know your thoughts!

mutdmour commented on packages/cli/commands/execute.ts:
```diff
@@ -163,24 +172,26 @@ export class Execute extends Command {
-			this.log('Execution was successfull:');
-			this.log('====================================');
-			this.log(JSON.stringify(data, null, 2));
+			console.info('Execution was successfull:');
```
nitpick - typo
```suggestion
			console.info('Execution was successful:');
```

mutdmour commented on packages/cli/src/Logger.ts:
```diff
@@ -0,0 +1,77 @@
+				winston.format.json()
+			);
+			this.logger.add(
+				new winston.transports.File({
+					filename: config.get('logs.file.location'),
+					format: fileLogFormat,
```
should not we add some form of log rotation? or is that handled elsewhere?
file transport format supports maxsize and maxFiles to support this https://github.com/winstonjs/winston/blob/master/docs/transports.md#file-transport

krynble replied:
Hi @mutdmour 

I am being very conservative to adding dependencies and it seems like adding log rotations would require another one.

The thoughts behind this first implementation is to create a basic logging framework that can be improved later, so for this reason I did not add any dependency for log output as all required new dependencies. Transports like elastic, loggly and cloudwatch logs all require more dependencies.

Maybe we can allow users to plug in other transports as they need. What do you think?

mutdmour replied:
adding maxsize and maxFiles should not require any new dependencies. File transport used here already supports these https://github.com/winstonjs/winston/blob/master/docs/transports.md#file-transport if I am not mistaken 
I have seen winston log files fill up disk before in a webapp without `maxsize` being set.

krynble replied:
Thanks for clarifying that @mutdmour .

I just realized that daily rotation would require a new lib but adding these settings do not.

Will apply those changes!

krynble committed:
Fix typos

krynble committed:
Adding basic file rotation to logs as suggested by @mutdmour

krynble committed:
Merge branch 'master' into logging

krynble committed:
Fixed linting issues

krynble committed:
Correcting comment to correctly reflect space usage

mutdmour commented on packages/cli/src/Logger.ts:
```diff
@@ -0,0 +1,79 @@
+				new winston.transports.File({
+					filename: config.get('logs.file.location'),
+					format: fileLogFormat,
+					maxsize: 16 * 1024 * 1024, // 16MB
+					maxFiles: 100, // 100 files of 16MB = 1.6GB
```
thanks for addressing this
one more thought - there might be best as configurable values rather than static

krynble replied:
Thanks again for your contribution @mutdmour 

Just updated with your suggestions!

krynble committed:
Added settings for log files rotation

mutdmour commented on packages/cli/config/index.ts:
```diff
@@ -566,6 +568,41 @@ const config = convict({
+				env: 'N8N_LOG_FILE_LOCATION',
+			},
+			maxFileSize: {
+				doc: 'Maximum size for each log file in MB.',
+				format: String,
+				default: '16',
+				env: 'N8N_LOG_FILE_MAXSIZE',
+			},
+			maxFileCount: {
+				doc: 'Maximum number of files to keep.',
+				format: String,
+				default: '100',
+				env: 'N8N_LOG_FILE_MAXCOUNT',
```
nitpick - format here can be Number

krynble replied:
Thanks again!

mutdmour reviewed (approved):
LGTM - just a small nitpick

krynble committed:
Correcting config type from String to Number

mutdmour commented on packages/cli/config/index.ts:
```diff
@@ -590,13 +590,13 @@ const config = convict({
 				doc: 'Maximum number of files to keep.',
-				format: String,
+				format: Number,
 				default: '100',
```
default values need to be updated too since type is now Number

krynble replied:
Again: thanks, again :D hahaha

mutdmour replied:
😊✅

krynble committed:
Changed default file settings to number

To reflect previous changes to the type

ahsanv reviewed (approved):
👍 Great work!

ivov commented on packages/cli/src/Logger.ts:
```diff
@@ -0,0 +1,79 @@
+	LogTypes,
+} from 'n8n-workflow';
+
+export class Logger implements ILogger {
```
To keep this as a singleton:
```suggestion
class Logger implements ILogger {
```

ivov commented on packages/workflow/src/LoggerProxy.ts:
```diff
@@ -0,0 +1,45 @@
+	instance = logger;
+}
+
+export function getInstance(): ILogger {
```
To make this more explicit than just "instance":

```suggestion
export function getLogger(): ILogger {
```

ivov commented on packages/workflow/src/LoggerProxy.ts:
```diff
@@ -0,0 +1,45 @@
+} from './Interfaces';
+
+
+let instance: ILogger | undefined;
+
+export function init(logger: ILogger) {
+	instance = logger;
```
To keep consistency with the getters:

```suggestion
let logger: ILogger | undefined;

export function init(loggerInstance: ILogger) {
	logger = loggerInstance;
```

ivov commented on packages/cli/src/ActiveWorkflowRunner.ts:
```diff
@@ -58,21 +60,27 @@ export class ActiveWorkflowRunner {
+		const logger = LoggerProxy.getInstance();
+
 		if (workflowsData.length !== 0) {
-			console.log('\n ================================');
-			console.log('   Start Active Workflows:');
-			console.log(' ================================');
+			logger.info('\n ================================');
+			logger.info('   Start Active Workflows:');
+			logger.info(' ================================');
```
When do we call the logger as `LoggerProxy.getInstance();` (see above) and when as `getLogger` (originally `getInstance`)? The interfaces are identical.

krynble replied:
Hello @ivov 

You should usually only use `LoggerProxy.getInstance()` as it should be available project-wide.

The `getLogger` function is used to initialize the logger for the first time and it is then provided to the `LoggerProxy`.

In a few cases, when the LoggerProxy has not been initialized yet you can use `getLogger` directly but it is recommended avoiding it as much as possible.

If it's unclear please let me know and I can clarify.

ivov commented on packages/cli/src/ActiveWorkflowRunner.ts:
```diff
@@ -88,6 +96,7 @@ export class ActiveWorkflowRunner {
 	 */
 	async removeAll(): Promise<void> {
 		const activeWorkflowId: string[] = [];
+		LoggerProxy.getInstance().verbose('Call to remove all active workflows received.');
```
Why not `LoggerProxy.verbose()` instead? Same for other examples in this file.

ivov commented on packages/cli/src/WebhookHelpers.ts:
```diff
@@ -30,13 +30,14 @@ import {
 	IWebhookData,
 	IWebhookResponseData,
 	IWorkflowExecuteAdditionalData,
+	LoggerProxy,
```
Maybe we do not need as much detail in the name when the logger is being used.

```suggestion
	LoggerProxy as Logger,
```

ivov commented on packages/cli/src/WorkflowExecuteAdditionalData.ts:
```diff
@@ -36,6 +36,7 @@ import {
 	IWorkflowExecuteAdditionalData,
 	IWorkflowExecuteHooks,
 	IWorkflowHooksOptionalParameters,
+	LoggerProxy,
```
For readability:

```suggestion
	LoggerProxy as Logger,
```

ivov commented on packages/cli/src/WorkflowRunner.ts:
```diff
@@ -29,7 +29,9 @@ import {
 import {
 	IDataObject,
 	IExecutionError,
+	ILogger,
```
Unused imports:

```suggestion
	IExecutionError,
```

ivov commented on packages/cli/src/WorkflowRunner.ts:
```diff
@@ -29,7 +29,9 @@ import {
 	IExecutionError,
+	ILogger,
 	IRun,
+	LoggerProxy,
```
```suggestion
	LoggerProxy as Logger,
```

ivov commented on packages/core/src/ActiveWorkflows.ts:
```diff
@@ -7,6 +7,7 @@ import {
 	IPollResponse,
 	ITriggerResponse,
 	IWorkflowExecuteAdditionalData,
+	LoggerProxy,
```
```suggestion
	LoggerProxy as Logger,
```

ivov commented on packages/workflow/src/LoggerProxy.ts:
```diff
@@ -0,0 +1,45 @@
+// Convenience methods below
+
+export function debug(message: string, meta: object = {}) {
+	getInstance().log('debug', message, meta);
```
Why not call it like this? Same for the other convenience methods.

```suggestion
	getLogger().debug(message, meta);
```

krynble replied:
I like the idea of calling the convenience methods statically. Changed the code so it looks like this:

```
import {
    LoggerProxy as Logger,
}  from 'n8n-workflow';

...

LoggerProxy.log(...);
LoggerProxy.error(...);
```

ivov reviewed (changes requested):
Omar, great job! We had a number of comments regarding importing and usage of the `Logger` vs. `LoggerProxy`, how to use them and the reason behind the convenience methods and when they are not used. Happy to jump on a quick call with you. 

Looking forward for the feature.

Done together with @BHesseldieck.

krynble committed:
Changed the way log messages are added to be called statically. Also minor naming improvements

krynble committed:
Merging master

krynble commented:
Thanks a lot @ivov @ahsanv @mutdmour and @BHesseldieck  for your contributions!

May I ask for one last look before we merge? I would also ask you to approve changes if it's all good =)

ivov commented on packages/cli/commands/execute.ts:
```diff
@@ -163,24 +172,26 @@ export class Execute extends Command {
-				this.log('Execution was NOT successfull:');
-				this.log('====================================');
-				this.log(JSON.stringify(data, null, 2));
+				console.info('Execution was NOT successful. See log message for details.');
+				logger.info('Execution error:');
```
Here as well, when do we use `console` and when `logger`?

ivov commented on packages/cli/commands/webhook.ts:
```diff
@@ -42,7 +49,7 @@ export class Webhook extends Command {
 	 */
 	static async stopProcess() {
-		console.log(`\nStopping n8n...`);
+		LoggerProxy.getInstance().info(`\nStopping n8n...`);
```
Can we do it like this? (whole file)

```suggestion
		LoggerProxy.info(`\nStopping n8n...`);
```

ivov commented on packages/cli/commands/worker.ts:
```diff
@@ -71,7 +79,7 @@ export class Worker extends Command {
 	 */
 	static async stopProcess() {
-		console.log(`\nStopping n8n...`);
+		LoggerProxy.getInstance().info(`\nStopping n8n...`);
```
Same as prior comment.

```suggestion
		LoggerProxy.info(`\nStopping n8n...`);
```

ivov commented on packages/cli/src/ActiveWorkflowRunner.ts:
```diff
@@ -60,20 +62,24 @@ export class ActiveWorkflowRunner {
-			console.log('\n ================================');
-			console.log('   Start Active Workflows:');
-			console.log(' ================================');
+			Logger.info('\n ================================');
+			Logger.info('   Start Active Workflows:');
+			Logger.info(' ================================');
```
Maybe it would be more appropriate for this to be `console.info()`?

ivov commented on packages/cli/src/Logger.ts:
```diff
@@ -0,0 +1,79 @@
+	
+	constructor() {
+
+		const logFormat = winston.format.printf(({ message }) => {
+			return message;
+		}) as winston.Logform.Format;
```
Maybe:

```suggestion
		const logFormat = winston.format.printf(({ message }) => message) as winston.Logform.Format;
```

ivov reviewed (approved):
Thanks Omar! Some minor remarks and suggestions.

krynble committed:
Applying latest corrections sent by @ivov

janober committed:
:twisted_rightwards_arrows: Merge branch 'master' into logging

janober committed:
:zap: Some logging improvements

krynble committed:
Saving logs to a folder inside n8n home instead of root

krynble committed:
Fixed broken tests and linting

krynble committed:
Changed some log messages to improve formatting

krynble committed:
Adding quotes to names  on log messages

mutdmour commented on packages/cli/commands/worker.ts:
```diff
@@ -79,7 +79,7 @@ export class Worker extends Command {
 	 */
 	static async stopProcess() {
-		LoggerProxy.getInstance().info(`\nStopping n8n...`);
+		LoggerProxy.info(`\nStopping n8n...`);
```
is this `\n` needed?

mutdmour commented on packages/cli/src/ActiveWorkflowRunner.ts:
```diff
@@ -61,27 +61,25 @@ export class ActiveWorkflowRunner {
-			logger.info('\n ================================');
-			logger.info('   Start Active Workflows:');
-			logger.info(' ================================');
+			console.info('\n ================================');
```
is this `\n` needed?

mutdmour commented on packages/cli/src/WorkflowExecuteAdditionalData.ts:
```diff
@@ -132,7 +132,7 @@ function hookFunctionsPush(): IWorkflowExecuteHooks {
 					return;
 				}
-				LoggerProxy.verbose(`Executing hook nodeExecuteBefore for hookFunctionsPush on node ${nodeName}`, {executionId: this.executionId});
+				Logger.verbose(`Executing hook on node "${nodeName}" (nodeExecuteBefore, hookFunctionsPush)`, {executionId: this.executionId});
```
why not include workflowId in logs in this file too? as well as sessionId? would make debugging that much easier.

mutdmour commented on packages/cli/src/Logger.ts:
```diff
@@ -1,31 +1,48 @@
+					winston.format.timestamp(),
+					winston.format.colorize({ all: true }),
+					winston.format.printf(({ level, message, timestamp, metadata }) => {
+						return `${timestamp} | ${level.padEnd(18)} | ${message}` + (Object.keys(metadata).length ? ` ${JSON.stringify(metadata)}` : '');
```
nitpick - spacing here seems off

krynble replied:
Hi @mutdmour 

I did not find any issue here, spacing seems fine (once before and after each pipe). Any suggestions?

mutdmour replied:
my bad. probably just looked funky on github

krynble committed:
Added execution and session IDs to logs. Also removed unnecessary line breaks

krynble committed:
Merge latest changes from master

krynble committed:
:zap: Added file caller to log messages (#1657)

This is done using callsites library which already existed
in the project as another library's dependency. So in fact
it does not add any new dependency.

mutdmour commented on packages/cli/src/Logger.ts:
```diff
@@ -57,29 +60,37 @@ class Logger implements ILogger {
+		// [2]: Should point to the caller.
+		// Note: getting line number is useless because at this point
+		// We are in runtime, so it means we are looking at compiled js files
+		const logDetails = callsite[2] === undefined ? {} : {file: basename(callsite[2].getFileName() || '')};
```
😊

krynble committed:
Adding logs to help debug Salesforce node

janober committed:
:twisted_rightwards_arrows: Merge branch 'master' into logging

janober committed:
:zap: Add function name to logs and add more logs

janober committed:
:zap: Improve some error messages

janober committed:
:zap: Improve some more log messages

janober committed:
:zap: Rename logging env variables to match others

janober commented:
Thanks a lot. Got merged.

janober commented:
Got released with n8n@0.118.0
